Feat: user configurable `draw.poses` colors

gao-hongnan opened the pull request:
This pull request fixes https://github.com/aisingapore/PeekingDuck-Private/issues/89.

- Allow users to configure `keypoint_dot_color` and `keypoint_connect_color`. Configurations for colors can be either a `str` or a `list` of BGR values. See example below:

    ```
    keypoint_dot_color: tomato # [77, 103, 255]
    keypoint_connect_color: champagne # [156, 223, 244]
    ```
- Refactor `draw.poses` Node to take in `ThresholdCheckerMixin` to validate configurations.
- Refactor `draw.utils.pose` such that drawing functions are grouped under a class named `Pose`.
- Add color palette under docstrings.

gao-hongnan committed:
feat: add user configurable `draw.poses` colors

gao-hongnan committed:
refactor: package functions inside pose.py as a class

gao-hongnan committed:
feat: add a comprehensive COLOR_MAP (color palette) to nodes.draw.utils.constants.py

gao-hongnan committed:
feat: implement function to convert input str or list to a tuple of BGR values

gao-hongnan committed:
refactor: move threshold and config checking out from utils/pose.py

gao-hongnan committed:
refactor: add threshold and config checking in draw.poses node

gao-hongnan committed:
docs: add two images to assets for render in docstrings of node draw.poses

gao-hongnan committed:
docs: update docstrings to include color palette for users to choose from

gao-hongnan committed:
refactor: set default colors in poses.yml to be in string instead of BGR values

gao-hongnan committed:
refactor: made logical flow changes to `_validate_configs` and `_check_valid_color`

gao-hongnan committed:
fix: fix mypy error

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -44,14 +47,95 @@ class Node(AbstractNode):
+       :header-rows: 1
+
+       * - Color Palette by Type
+         - Color Palette by Name
```
Would "Color Palette by Hue" be a better description of how the colors are sorted?

gao-hongnan replied:
Yeap, agree on this and changed accordingly.

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -44,14 +47,95 @@ class Node(AbstractNode):
+
+           PeekingDuck's Color Palette Sorted by Color Types [1]_.
+
+         - .. figure:: ../assets/api/color_map_by_name.png
```
Please add the code used to generate these images in `scripts/docs`

gao-hongnan replied:
Added

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -44,14 +47,95 @@ class Node(AbstractNode):
+
+        keypoint_connect_color (:obj:`Union[List[int], str]`): **default = "champagne"** |br|
+            Color of the keypoints should either be a string in :ref:`color-palette`,
+            or a list of BGR values.
```
I think this should say "Color of the keypoint connections..." instead

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -44,14 +47,95 @@ class Node(AbstractNode):
+        self.check_bounds("keypoint_dot_radius", "[0, +inf)")
+
+    def _check_valid_color(self, key: str) -> None:
+        """Checks that configuration value specified by `key` is a valid color.
```
Place private methods below public method and arrange alphabetically

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -44,14 +47,95 @@ class Node(AbstractNode):
+                values are out of range [0, 255].
+        """
+        valid_colors_type = set(COLOR_MAP.keys())
+        valid_colors_range = "[0, 255]"
```
I think these should be named as 

`valid_color_names` since it's a set containing multiple color names.

`valid_color_range` since every where else in this code uses `color`

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -44,14 +47,95 @@ class Node(AbstractNode):
-        self.keypoint_text_color = tuple(self.config["keypoint_text_color"])
+        self._validate_configs()
+
+        self.keypoint_dot_color = get_color(self.config["keypoint_dot_color"])
+        self.keypoint_connect_color = get_color(self.config["keypoint_connect_color"])
+
```
Consider using the attributes `keypoint_dot_color` and `keypoint_connect_color` directly instead of accessing them through `self.config`

gao-hongnan replied:
For the time being I revert back to `self.config["keypoint_dot_color"]` as `mypy` throwing `has-type` error:

```
Cannot determine type of "keypoint_dot_color"
```

liyier90 replied:
Just another thought, is there a need to reassign `self.keypoint_dot_color = get_color(...)` instead of just passing it directly to `Pose`'s constructor? Eg:

```
self.pose = Pose(
    get_color(self.keypoint_dot_color),
    ...
)
```

gao-hongnan replied:
I think that's ok too, I can change it accordingly.

liyier90 commented on peekingduck/pipeline/nodes/draw/utils/constants.py:
```diff
@@ -148,3 +148,160 @@
+    "blanchedalmond": (205, 235, 255),
+    "blue": (255, 0, 0),
+    "blueviolet": (226, 43, 138),
+    "brown": (42, 42, 165),
```
I think there is a peekingduck brown (96, 109, 167), consider adding that in as `pkdbrown` to avoid name clashing.

gao-hongnan replied:
Added in PeekingDuck's brown as `pkdbrown`.

liyier90 commented on peekingduck/pipeline/nodes/draw/utils/general.py:
```diff
@@ -61,3 +63,25 @@ def project_points_onto_original_image(
+
+    Args:
+        color (Union[str, List[int]]): Color can be a string from COLOR_MAP defined
+            in ./constants.py or a list of 3 integers representing BGR values.
```
I don't think the "defined in ./constants.py" is necessary since it's "taken care of" by the import statement

liyier90 commented on peekingduck/pipeline/nodes/draw/utils/pose.py:
```diff
@@ -13,80 +13,117 @@
+            keypoint_connections (Union[None, Iterable[Any]]): Keypoint connections
+                of one individual human with shape (15, 2, 2).
+            image_size (Tuple[int, int]): Image size of the input image frame.
+        """
```
Can type `keypoint_connections` as `Optional[Iterable[Any]]` to be consistent with how we type other `None | SomeType` variables

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -44,14 +47,95 @@ class Node(AbstractNode):
+                in :ref:`color-palette`, or a list of BGR values, or if the BGR
+                values are out of range [0, 255].
+        """
+        valid_colors_type = set(COLOR_MAP.keys())
```
Apply `sorted()` to this so the error message later reads better

gao-hongnan replied:
As discussed, will apply `sorted` in the error message instead.

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -44,14 +47,95 @@ class Node(AbstractNode):
+                raise ValueError(
+                    f"{key} must be one of {valid_colors_type} or passed as a list of BGR values."
+                ) from wrong_color_choice
+        else:
+            self.check_bounds(key, valid_colors_range)
```
Include check for the length of BGR list as well. So we throw better error message as well.

I can run the code with `keypoint_dot_color: [1, 2, 3, 4]` with no error.

And the error thrown for `keypoint_dot_color: [1, 2]` is quite cryptic:

```
   File "./peekingduck/pipeline/nodes/draw/utils/general.py", line 86, in get_color
    color_tuple = (color[0], color[1], color[2])

2022-10-28 08:36:51 root  ERROR:  IndexError: list index out of range
```

gao-hongnan replied:
Good catch, will change.

liyier90 reviewed (changes requested):
Some comments for your consideration. 

Please update the unit test for poses as well at `tests/pipeline/nodes/draw/test_poses.py`

gao-hongnan committed:
docs: this commit addresses pull request #717 and mainly fixes docstrings

gao-hongnan commented on peekingduck/pipeline/nodes/draw/utils/pose.py:
```diff
@@ -13,80 +13,117 @@
+        """
+        image_size = get_image_size(image)
+        num_persons = all_keypoints.shape[0]
+        if num_persons > 0:
+            for keypoints, keypoint_connections in zip(
+                all_keypoints, all_keypoint_connections
+            ):
```
As noted, will also benchmark on the performance of `zip` versus indexing.

gao-hongnan replied:
Benchmarked, `zip` seems to have similar performance.

gao-hongnan committed:
refactor: this commit addresses pull request #717 and mainly include minor code clean up, including
adding in a method `_check_valid_length` to check whether the length of the input is valid

gao-hongnan committed:
fix: fix line-too-long error

gao-hongnan committed:
fix: fix has-type error in mypy

gao-hongnan committed:
refactor/test: rename `TestBtmMidpoint` to `TestPoses`

gao-hongnan committed:
refactor/test: initialize empty keypoints, scores and conns with the correct empty shape in `test_no_poses`

gao-hongnan committed:
script: add script to generate color palette

gao-hongnan committed:
test: add tests for draw.poses

gao-hongnan committed:
test: add mock unit test in `test_poses` to test if color configs are indeed applied to the pose

liyier90 commented on peekingduck/pipeline/nodes/draw/poses.py:
```diff
@@ -60,5 +103,75 @@ def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+                    list of BGR values."
+                ) from wrong_color_choice
+        else:
+            self._check_valid_length(key)
+            self.check_bounds(key, valid_color_range)
```
Check valid length seems vague here, and doesn't really inform us what we consider to be a valid length.

Since we already know we're dealing with a list contain BGR value by the time we reach `else`, consider doing both the length and individual color channel value checks in a `_check_valid_bgr_value()` method. Example:

```
def _check_valid_bgr_value(self, key, color_range):
    if len(self.config[key]) != 3:
        raise ...
    self.check_bounds(key, valid_color_range)
```
I think it's fine to use a magic number `3` there since we already specify in the method name that we're checking BGR values

gao-hongnan replied:
Changed to check both length and bounds.

liyier90 commented on scripts/docs/generate_color_palette.py:
```diff
@@ -0,0 +1,268 @@
+from typing import Tuple
+
+import matplotlib.colors as mcolors
+import matplotlib.pyplot as plt
+from matplotlib.patches import Rectangle
+from PIL import ImageColor
```
Would be good to include a module docstring which says that this script is used to generate the color palette figures for the documentation

gao-hongnan replied:
Changed.

liyier90 commented on tests/pipeline/nodes/draw/test_poses.py:
```diff
@@ -16,39 +16,147 @@
+
+    @mock.patch("cv2.line", wraps=cv2.line)
+    @mock.patch("cv2.circle", wraps=cv2.circle)
+    def test_if_configs_are_used(
+        self, mock_circle, mock_line, draw_poses_config, create_image
+    ):
+        draw_poses_node = Node(draw_poses_config)
+
+        default_keypoint_dot_color = draw_poses_node.keypoint_dot_color
+        default_keypoint_connect_color = draw_poses_node.keypoint_connect_color
```
I think it would be better to set a custom color for this test, even something like `[1, 2, 3]` would be fine. Just to ensure that the values from the configs are passed down to the actual draw functions instead of relying on any hardcoded values.

gao-hongnan replied:
Thanks for pointing out, I have changed accordingly.

liyier90 reviewed (changes requested):
Just a few final minor changes. 

I notice there a some unaddressed comments from the previous round of review. Are these missed by accident or are there no actions/changes necessary for those? 

It would be good to leave a reply so it's easier to keep track of what's addressed and what isn't, even something as simple as "Change." is fine.

gao-hongnan commented:
> Just a few final minor changes.
> 
> I notice there a some unaddressed comments from the previous round of review. Are these missed by accident or are there no actions/changes necessary for those?
> 
> It would be good to leave a reply so it's easier to keep track of what's addressed and what isn't, even something as simple as "Change." is fine.

Apologies, I missed out on a few comments after making some changes. Will take note on this.

gao-hongnan committed:
refactor: pass init attributes to `Pose` constructor directly

gao-hongnan committed:
refactor: rename `_check_valid_length` to `_check_valid_bgr_value` as the latter is more explicit about what the function does
also combined `check_bounds` into this method

gao-hongnan committed:
script: add module docstring to scripts/docs/generate_color_palette.py

gao-hongnan committed:
test/refactor: preliminary change on `test_if_configs_are_used`

gao-hongnan committed:
test: modify test logic in `test_if_configs_are_used` to be more robust
